### PR TITLE
linkcheck: Update configuration defaults for Sphinx 8.0

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Refactoring

### Purpose
- Updates some `linkcheck` builder configuration defaults as-communicated by the deprecation process for the Sphinx v8.0 release.

### Detail
- Links that respond with HTTP 401 (unauthorized) responses are now considered `broken` by default.
- Timeouts that occur when checking a link are now reported with a distinct `timeout` status code, instead of the previous `broken`.
- The previous behaviours are still available and can be configured on an opt-in basis per-project using `conf.py`.

### Relates
- Relates to #12629 
- Relates to #11433 
- Relates to #11876

### CODE REVIEW

**Comment by jayaddison** [comment]
Ah; needs a `CHANGES.rst` entry (and awaits 8.x changelog heading).  Adding do-not-merge label..

File: sphinx/builders/linkcheck.py [hunk 1]
```diff
@@ -717,8 +675,8 @@ def setup(app: Sphinx) -> ExtensionMetadata:
     app.add_config_value('linkcheck_anchors_ignore', ['^!'], '')
     app.add_config_value('linkcheck_anchors_ignore_for_url', (), '', (tuple, list))
     app.add_config_value('linkcheck_rate_limit_timeout', 300.0, '', (int, float))
-    app.add_config_value('linkcheck_allow_unauthorized', True, '')
-    app.add_config_value('linkcheck_report_timeouts_as_broken', True, '', bool)
+    app.add_config_value('linkcheck_allow_unauthorized', False, '', bool)
```
Comment (jayaddison) [inline sphinx/builders/linkcheck.py]:
```suggestion
    app.add_config_value('linkcheck_allow_unauthorized', False, '')
```

I had snuck in a type definition change here, but let's follow best-practice and not do that with unrelated changes.

File: CHANGES.rst [hunk 2]
```diff
@@ -9,6 +9,8 @@ Incompatible changes
 
 Deprecated
 ----------
+* #12630: Configuration defaults for ``linkcheck_allow_unauthorized`` and
+  ``linkcheck_report_timeouts_as_broken`` have been updated.
```
Comment (AA-Turner) [inline CHANGES.rst]: Is this the right section? Or do you also intend to deprecate the config values themselves?

A

Comment (jayaddison) [reply CHANGES.rst]: I did wonder about that; it's (hopefully!) the end of the deprecation process for these, not the start of one -- but I did think it'd be worth mentioning in the changelog for upgraders.  Is there a better place to put them?  (none of the other categories initially seem suitable)

**Comment by AA-Turner** [comment]
I've added a draft of the changes to configuration.rst -- feel free to rewrite.

A

File: CHANGES.rst [hunk 3]
```diff
@@ -9,6 +9,10 @@ Incompatible changes
 
 Deprecated
 ----------
+* #12630: Two ``linkcheck`` defaults are updated from deprecated legacy values:
```
Comment (jayaddison) [inline CHANGES.rst]: This line still reads a bit awkwardly to me.

File: CHANGES.rst (hunk 2 above)
Comment (AA-Turner) [reply CHANGES.rst]: Technically, *Incompatible changes*, as this is a backwards-incompatible change.

File: CHANGES.rst (hunk 3 above)
Comment (AA-Turner) [reply CHANGES.rst]: How about "Sphinx 8 enables two recommended linkcheck settings by default:", or a words to that effect?

File: CHANGES.rst (hunk 2 above)
Comment (jayaddison) [reply CHANGES.rst]: Mmm; I'm on the fence but tempted to leave it in _Deprecated_ since it is deprecation-related.

Comment (AA-Turner) [reply CHANGES.rst]: I view *Deprecated* as for new deprecations -- this is relating to an old deprecation. Note that e.g. changes in APIs go into *Incompatible changes* rather than *Deprecated*, when the deprecated behaviour gets removed.

A

Comment (jayaddison) [reply CHANGES.rst]: Ok.  I suppose theoretically the `timeout` status could cause code that's unaware of it to break.

Comment (jayaddison) [reply CHANGES.rst]: And I'm glad I re-read the entry as a result of that, because I've mis-stated the updated default.

**Review by AA-Turner — APPROVED** [review]

**Comment by AA-Turner** [comment]
Thanks!

A

**Comment by jayaddison** [comment]
Thank you!
